modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -0,0 +1,560 @@
+
+                    boolean value;
+                    if (negate) {
+                        // NOT (col IS FALSE) => col IS TRUE => col = true
+                        value = node.getKind() == SqlKind.IS_FALSE;
+                    } else {
+                        // NOT (col IS TRUE) => col IS FALSE => col = false
+                        value = node.getKind() == SqlKind.IS_TRUE;
+                    }
+
+                    return new PruningColumnSet(ref.getIndex(), rexBuilder.makeLiteral(value));
+                } else if (operand.isA(SqlKind.LOCAL_REF)) {
+                    return Result.RESTRICT;
+                } else {
+                    return Result.UNKNOWN;
+                }
+            }
+            case IS_NOT_FALSE:
+            case IS_NOT_TRUE: {
+                // IS_NOT_TRUE is used by case/when expression rewriter.
+                // IS_NOT_FALSE is added for symmetry.
+
+                boolean value;
+                if (negate) {
+                    value = node.getKind() != SqlKind.IS_NOT_FALSE;
+                } else {
+                    value = node.getKind() != SqlKind.IS_NOT_TRUE;
+                }
+
+                if (isColocationKey(operands.get(0), keys)) {
+                    RexLocalRef column = (RexLocalRef) operands.get(0);
+                    return new PruningColumnSet(column.getIndex(), rexBuilder.makeLiteral(value));
+                } else {
+                    return extractMetadata(operands.get(0), keys, rexBuilder, !negate);
+                }
+            }
+            default: {
+                if (node.isA(SqlKind.BINARY_COMPARISON)) {
+                    // Convert binary comparision operations to Result::RESTRICT
+                    RexNode lhs;
+                    RexNode rhs;
+
+                    if (operands.get(0).isA(SqlKind.LOCAL_REF)) {
+                        lhs = operands.get(0);
+                        rhs = operands.get(1);
+                    } else {
+                        lhs = operands.get(1);
+                        rhs = operands.get(0);
+                    }
+                    if (isColocationKey(lhs, keys) && isValueExpr(rhs)) {
+                        // We can not extract values from expressions such as colo_key > 10
+                        return Result.UNKNOWN;
+                    } else if (lhs.isA(SqlKind.LOCAL_REF) && isValueExpr(rhs)) {
+                        // We can use non_colo_key > 10 to narrow the search space.
+                        return Result.RESTRICT;
+                    } else {
+                        return Result.UNKNOWN;
+                    }
+                } else {
+                    return Result.UNKNOWN;
+                }
+            }
+        }
+    }
+
+    private static boolean isColocationKey(RexNode node, List<Integer> keys) {
+        if (node instanceof RexLocalRef) {
+            RexLocalRef localRef = (RexLocalRef) node;
+            return keys.contains(localRef.getIndex());
```

Comment:
`contains` use unboxed value.
Maybe it make sense to use `ImmutableIntList` interface and `ImmutableIntList.indexOf(int)` instead. Or use fastutils`IntList` and rely data  from Calcite immutable list, right after getting from `distribution` via `IntArrayList.of(ImmutableIntList.toArray())`

Reply:
ImmutableIntList.contains has complexity O(n) as i can see.

Reply:
I believe there will be few int values (less then 10). 
So, complexity may affect less than data locality, bucket calculation and additional level of indirection.